spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java

```diff
@@ -875,7 +875,13 @@ private static void deleteManifests(FileIO io, List<ManifestFile> manifests) {
   public static Dataset<Row> loadTable(SparkSession spark, Table table, long snapshotId) {
     SparkTable sparkTable = new SparkTable(table, snapshotId, false);
     DataSourceV2Relation relation = createRelation(sparkTable, ImmutableMap.of());
-    return Dataset.ofRows(spark, relation);
+    if (!(spark instanceof org.apache.spark.sql.classic.SparkSession)) {
```

Comment:
Spark 4.0 has [abstract class SparkSession](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala#L63) and [classic.SparkSession](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala#L63). We will always try to use the general one, and only cast to `classic.SparkSession` when the private APIs are needed.

The same rule applies to `Dataset`

Reply:
I think I noted other places, but let's change all of these checks to a Precondition if we can